Core/CollectionUtility.cs

```diff
@@ -0,0 +1,57 @@
+// This file is part of the re-linq project (relinq.codeplex.com)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+//
+// re-linq is free software; you can redistribute it and/or modify it under
+// the terms of the GNU Lesser General Public License as published by the
+// Free Software Foundation; either version 2.1 of the License,
+// or (at your option) any later version.
+//
+// re-linq is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-linq; if not, see http://www.gnu.org/licenses.
+//
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Remotion.Linq.SqlBackend
+{
+  /// <summary>
+  /// Provides utility services for handling collections.
+  /// </summary>
+  public static class CollectionUtility
+  {
+    /// <summary>
+    /// Determines whether the given <paramref name="value"/> is a collection, i.e. one of the following:
+    /// <list>
+    ///   <item><term><see cref="ICollection"/></term></item>
+    ///   <item><term><see cref="ICollection{T}"/></term></item>
+    ///   <item><term><see cref="IReadOnlyCollection{T}"/></term></item>
+    /// </list>
+    /// </summary>
+    public static bool IsCollection (object value)
```

Comment:
Not sure what char[] would be used for, but it would seem reasonable to also consider it this way? @martin-herwich

Reply:
Added `string` for now.

Reply:
`string` would be handled correctly by the method as is, because `string` is neither an `ICollection`, nor `ICollection<T>`, nor `IReadOnlyCollection<T>`, but it may be more efficient to check it explicitly before we check for generic collections.

`char[]` would be detected as a collection, which may or may not be correct, but treating it as a collection is in line with the previous implementation - see `SqlPreparationExpressionVisitor.VisitConstant()`, which only tested for `ICollection`.
My intuition is that `char[]` should be treated as a collection by default; if we want to use `char[]` for an `AnsiString`/`varchar` parameter (as opposed to `string`, which is a `String`/`nvarchar`) in the context of `Remotion.Data`, we should override the default behavior there.
However, this seems impossible at the moment, because `QueryExecutor.GenerateSqlCommand()` has a hard-coded `new DefaultSqlPreparationStage()`, whose `PrepareExpression` hard-codes the usage of `SqlPreparationExpressionVisitor`, so there is no extension point.

Reply:
> string would be handled correctly by the method as is, because string is neither an ICollection, nor ICollection<T>, nor IReadOnlyCollection<T>,

Ah, that's true. I stopped understanding the code past the IF IS IEnumerable in terms of that string wouldn't be true. Yes, defintely more efficient. but mostly, easier to parse :)

Reply:
> My intuition is that char[] should be treated as a collection by default; if we want to use char[] for an AnsiString/varchar parameter (as opposed to string, which is a String/nvarchar) in the context of Remotion.Data, we should override the default behavior there.

I just checked SqlDataParameter and now I know where the `char[]` stuff comes from: SqlDataParameter interprets a `char[]` value as a `string` type. So, when we see it in re-linq SQL Backend (or in re-store) it means "scalar text" not "collection"

Reply:
`SqlParameter` does assume `dataType=typeof(string)` when the `DbType` _is not set_ and the value is a `char` or `char[]`. However, it does not follow that we should assume `char[]` is meant as a `string` when parsing the expression tree - we can make a `char[]` into a Table-Valued parameter.

So, one question is: what did the developer most likely intend by using a char[] instead of string when writing the LINQ query?
Plus: if a developer wants to use a char[] as a collection, what can he do if we interpret it as string?
In the inverted case, just use a string instead of a `char[]` if you want it interpreted as a scalar value  -simple and intuitive.

Furthermore, since this feature is about the `Contains`-operator, it does not make sense to assume that the `char[]` is supposed to be a `string`, because then the `WHERE IN`-clause would be broken.

Reply:
It says "The following table shows the inferred .NET Framework type", which means that if a `char`, `varchar`, `nchar`, or `nvarchar` value is read from the database, the .NET type is inferred to be either `string` or `char[]`, so that you can use both the `GetString()` and `GetChars()` methods on `SqlDataReader`, depending on whether the data is to be interpreted as a single string value or as an array of characters.

But this is all in the MSSQL-specific ADO.NET implementation, and has nothing to do with LINQ expressions. At this point in the code, we only care about whether the developer meant a single `string` value or a collection of `char`s when he used a `ConstantExpression` with a `char[]`.
Since we only have an MSSQL-specific implementation in our Sql-Backend, we could lean on the MSSQL-specific ADO.NET implementation to answer this question, but that may become problematic if we ever have to support other database systems with our Sql-Backend.

Note that interpreting `char[]` as a single value is also a breaking change.

Reply:
Given that both PostgreSQL and Orcale consider `char[]` to be text-types and the fact that until now, both `byte[]` and `char[]` where generated as individual parameters for each element in the array, supporting them still seems to be the least-suprising behavior - from the POV of ADO.NET. It's true that a c# developer might expect different behavior. However, given that `byte[]` is the only representation available for `binary` in SQL Server, it seems strange to handle `byte[]` and `char[]` differently. 

Neither solution is perfect but given that it comes down to only adding tis one condition and then having both scenarious handled identically in SqlBackend, this seems the best choice. One day, we might want to extend this via an SQL dialect to allow a choice of interpreting an array as a scalar type or a collection.

Reply:
byte[] and char[] are both treated as single values now; this is documented in https://re-motion.atlassian.net/browse/RMLNQSQL-176